x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java

```diff
@@ -105,18 +138,24 @@ public void putAsync(ILMHistoryItem item) {
                 LIFECYCLE_HISTORY_INDEX_ENABLED_SETTING.getKey(), item);
             return;
         }
-        logger.trace("about to index ILM history item in index [{}]: [{}]", ILM_HISTORY_ALIAS, item);
-        ensureHistoryIndex(client, clusterService.state(), ActionListener.wrap(createdIndex -> {
-            try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
-                item.toXContent(builder, ToXContent.EMPTY_PARAMS);
-                IndexRequest request = new IndexRequest(ILM_HISTORY_ALIAS).source(builder);
-                processor.add(request);
-            } catch (IOException exception) {
-                logger.error(new ParameterizedMessage("failed to index ILM history item in index [{}]: [{}]",
-                    ILM_HISTORY_ALIAS, item), exception);
-            }
-        }, ex -> logger.error(new ParameterizedMessage("failed to ensure ILM history index exists, not indexing history item [{}]",
-            item), ex)));
+        logger.trace("queueing ILM history item for indexing [{}]: [{}]", ILM_HISTORY_ALIAS, item);
+        try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+            item.toXContent(builder, ToXContent.EMPTY_PARAMS);
+            IndexRequest request = new IndexRequest(ILM_HISTORY_ALIAS).source(builder);
+            // TODO: remove the threadpool wrapping when the .add call is non-blocking
+            //  (it can currently execute the bulk request occasionally)
+            threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
```

Comment:
shall we grab the executor from the `threadpool` in the constructor to avoid this lookup on every `putAsync` ?

Reply:
This should be relatively cheap since it is a map lookup. In addition, we should look it up every time in the event that the executor changes (configuration changes used to change executors if I recall correctly)